helix-core/src/main/java/org/apache/helix/controller/dataproviders/ResourceControllerDataProvider.java

```diff
@@ -583,6 +586,37 @@ public Set<CapacityNode> getSimpleCapacitySet() {
     return _simpleCapacitySet;
   }
 
+  public void populateSimpleCapacitySetUsage(final Set<String> resourceNameSet,
+      final CurrentStateOutput currentStateOutput) {
+    // Convert the assignableNodes to map for quick lookup
+    Map<String, CapacityNode> simpleCapacityMap = new HashMap<>();
+    for (CapacityNode node : _simpleCapacitySet) {
+      simpleCapacityMap.put(node.getId(), node);
+    }
+    for (String resourceName : resourceNameSet) {
+      // Process current state mapping
+      for (Map.Entry<Partition, Map<String, String>> entry : currentStateOutput.getCurrentStateMap(
+          resourceName).entrySet()) {
+        for (String instanceName : entry.getValue().keySet()) {
+          CapacityNode node = simpleCapacityMap.get(instanceName);
+          if (node != null) {
+            node.canAdd(resourceName, entry.getKey().getPartitionName());
+          }
+        }
+      }
+      // Process pending state mapping
+      for (Map.Entry<Partition, Map<String, Message>> entry : currentStateOutput.getPendingMessageMap(
+          resourceName).entrySet()) {
+        for (String instanceName : entry.getValue().keySet()) {
+          CapacityNode node = simpleCapacityMap.get(instanceName);
+          if (node != null) {
+            node.canAdd(resourceName, entry.getKey().getPartitionName());
+          }
+        }
+      }
```

Comment:
This code is redundant.

You can have a private function with argument for the map to process with same logic and hook twice here.